torch/csrc/api/src/optim/lbfgs.cpp

```diff
@@ -10,167 +10,609 @@
 #include <cmath>
 #include <functional>
 #include <vector>
+#include <algorithm>
 
 namespace torch {
 namespace optim {
 
-LBFGSOptions::LBFGSOptions(double learning_rate)
-    : learning_rate_(learning_rate) {}
+LBFGSOptions::LBFGSOptions(double lr) : lr_(lr) {}
 
-Tensor LBFGS::gather_flat_grad() {
+bool operator==(const LBFGSOptions& lhs, const LBFGSOptions& rhs) {
+  return (lhs.lr() == rhs.lr()) &&
+         (lhs.max_iter() == rhs.max_iter()) &&
+         (lhs.max_eval() == rhs.max_eval()) &&
+         (lhs.tolerance_grad() == rhs.tolerance_grad()) &&
+         (lhs.tolerance_change() == rhs.tolerance_change() &&
+         (lhs.history_size() == rhs.history_size())) &&
+         (lhs.line_search_fn() == rhs.line_search_fn());
+}
+
+void LBFGSOptions::serialize(torch::serialize::OutputArchive& archive) const {
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(lr);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(max_iter);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(max_eval);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(tolerance_grad);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(tolerance_change);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(history_size);
+  _TORCH_OPTIM_SERIALIZE_TORCH_ARG(line_search_fn);
+}
+
+void LBFGSOptions::serialize(torch::serialize::InputArchive& archive) {
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, lr);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, max_iter);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG_OPTIONAL(int64_t, max_eval);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, tolerance_grad);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(double, tolerance_change);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG(int64_t, history_size);
+  _TORCH_OPTIM_DESERIALIZE_TORCH_ARG_OPTIONAL(std::string, line_search_fn);
+}
+
+template <typename T>
+bool if_container_equal(T lhs, T rhs) {
+  if (!(lhs.size() == rhs.size())) return false;
+  for (size_t i = 0; i < lhs.size(); i++) {
+    if (!torch::equal(lhs.at(i), rhs.at(i))) return false;
+  }
+  return true;
+}
+
+bool operator==(const LBFGSParamState& lhs, const LBFGSParamState& rhs) {
+  auto isNull = [](c10::optional<std::vector<Tensor>> val) { return val == c10::nullopt; };
+  return (lhs.func_evals() == rhs.func_evals()) &&
+         (lhs.n_iter() == rhs.n_iter()) &&
+         (lhs.t() == rhs.t()) &&
+         (lhs.prev_loss() == rhs.prev_loss()) &&
+         torch::equal_if_defined(lhs.d(), rhs.d()) &&
+         torch::equal_if_defined(lhs.H_diag(), rhs.H_diag()) &&
+         torch::equal_if_defined(lhs.prev_flat_grad(), rhs.prev_flat_grad()) &&
+         if_container_equal(lhs.old_dirs(), rhs.old_dirs()) &&
+         if_container_equal(lhs.old_stps(), rhs.old_stps()) &&
+         if_container_equal(lhs.ro(), rhs.ro()) &&
+         ((isNull(lhs.al()) && isNull(rhs.al())) ||
+            (!isNull(lhs.al()) && !isNull(rhs.al()) && if_container_equal(*lhs.al(), *rhs.al())));
```

Comment:
I think `lhs.al() == rhs.al()` should just work for this use case :D https://github.com/pytorch/pytorch/blob/a4224886f3e03991161d077bac7bedbdac534170/c10/util/Optional.h#L631-L634